source/kernel/src/serial_trace.rs

```diff
@@ -378,10 +359,94 @@ impl Subscriber for SerialSubscriber {
     }
 }
 
+use serde::{Deserialize, Serialize};
+
+use crate::services;
+
+#[derive(Debug, Serialize, Deserialize)]
+#[non_exhaustive]
+pub struct SerialTraceSettings {
+    /// SerialMux port for sermux tracing.
+    pub port: u16,
+
+    /// Capacity for the serial port's send buffer.
+    pub sendbuf_capacity: usize,
+
+    /// Capacity for the trace ring buffer.
+    ///
+    /// Note that *two* buffers of this size will be allocated. One buffer is
+    /// used for the normal trace ring buffer, and another is used for the
+    /// interrupt service routine trace ring buffer.
+    pub tracebuf_capacity: usize,
+
+    /// Initial level filter used if the debug host does not select a max level.
+    #[serde(with = "level_filter")]
+    pub initial_level: tracing::metadata::LevelFilter,
+}
+
+pub const fn level_to_u8(level: tracing::metadata::LevelFilter) -> u8 {
+    match level {
+        tracing::metadata::LevelFilter::TRACE => 0,
+        tracing::metadata::LevelFilter::DEBUG => 1,
+        tracing::metadata::LevelFilter::INFO => 2,
+        tracing::metadata::LevelFilter::WARN => 3,
+        tracing::metadata::LevelFilter::ERROR => 4,
+        tracing::metadata::LevelFilter::OFF => 5,
+    }
+}
+
+pub const fn u8_to_level(level: u8) -> tracing::metadata::LevelFilter {
+    match level {
+        0 => tracing::metadata::LevelFilter::TRACE,
+        1 => tracing::metadata::LevelFilter::DEBUG,
+        2 => tracing::metadata::LevelFilter::INFO,
+        3 => tracing::metadata::LevelFilter::WARN,
+        4 => tracing::metadata::LevelFilter::ERROR,
+        _ => tracing::metadata::LevelFilter::OFF,
+    }
+}
+
+mod level_filter {
+    use serde::{de::Visitor, Deserializer, Serializer};
+
+    use super::{level_to_u8, u8_to_level};
+
+    pub fn serialize<S>(lf: &tracing::metadata::LevelFilter, s: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let lf_u8: u8 = level_to_u8(*lf);
+        s.serialize_u8(lf_u8)
+    }
+
+    pub fn deserialize<'de, D>(d: D) -> Result<tracing::metadata::LevelFilter, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct LFVisitor;
+        impl<'de> Visitor<'de> for LFVisitor {
+            type Value = tracing::metadata::LevelFilter;
+
+            fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+                formatter.write_str("a level filter as a u8 value")
+            }
+
+            fn visit_u8<E>(self, v: u8) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(u8_to_level(v))
+            }
+        }
+
+        d.deserialize_u8(LFVisitor)
+    }
+}
```

Comment:
hmm...i think representing levels as integers in the SerMux wire proto makes sense, but i don't know if I like representing them as integers in the config file. my preference would be to use [the `FromStr` impl for `LevelFilter`](https://docs.rs/tracing/latest/tracing/level_filters/struct.LevelFilter.html#impl-FromStr-for-LevelFilter), so that level filters in a config file are specified as, e.g.
```toml
[serial_trace]
initial_level = "debug"
```
rather than
```toml
[serial_trace]
initial_level = 1
```

Reply:
Ahhh, you're probably right. I had my "postcard" brain on, but strings are probably better because this will round trip through json/toml too.